### ISSUE TITLE
自己数据集训练的yolov9模型tensorrt的输出结果与pytorch结果差别很大，精度完全没有对齐

### ISSUE THREAD

**Issue by guanshanjushib** [wang-xinyu/tensorrtx#1535]
## Env

- GPU, e.g. RTX3060
- OS, e.g. Ubuntu20.04
- Cuda version: cuda-11.1
- TensorRT version: tensorrt-7.2.3

## About this repo

- which branch/tag/commit are you using？
master-branch
- which model? 
yolov9

## Your problem
我用yolov9算法训练自己的数据集，得到的模型经过yolov9的tensorrtx推理，发现精度对不齐，pytorch输出的结果与tensorrt输出的结果差别很大，有些目标框的位置和精度没对齐。请问当前的yolov9推理对齐了吗？

**Comment by guanshanjushib** [comment]
请问有人训练自己的数据集后pytorch和tensorrt的结果对齐了吗

**Comment by wang-xinyu** [comment]
@guanshanjushib 试试用官方的模型
cc @WuxinrongY

**Comment by WuxinrongY** [comment]
你能发一下模型的结构吗，在pytorch下print打印就可以。如果和官网的一致，可以考虑提高一下tensorrt的版本，采用trt 8.6。

**Comment by guanshanjushib** [comment]
> 你能发一下模型的结构吗，在pytorch下print打印就可以。如果和官网的一致，可以考虑提高一下tensorrt的版本，采用trt 8.6。

[yolov9_uav76_20240521_key.txt](https://github.com/user-attachments/files/15585207/yolov9_uav76_20240521_key.txt)
[yolov9_uav76_20240521_model.txt](https://github.com/user-attachments/files/15585209/yolov9_uav76_20240521_model.txt)
附件里是我生成wts后保存的模型结构和wts每层的key和len，我采用的是yolov9-e模型，用的是yolov9的master分支训练的

**Comment by WuxinrongY** [comment]
你可以关注一下SyncBatchNorm的实现，是不是和官方模型中的BatchNorm2d不一样，可以直接在代码里面修改。maxOutput这个不要紧的，输出的是最大的置信度，至于预处理、网络结构和后处理方面确实可以讨论一下，hhh
![image](https://github.com/wang-xinyu/tensorrtx/assets/53141838/86e3c5b3-095a-45b6-90f8-ef5a0eff8aff)

**Comment by WuxinrongY** [comment]
要是有兴趣，我们可以考虑实现一下SyncBatchNorm。email: w845756@qq.com

**Comment by guanshanjushib** [comment]
> 你可以关注一下SyncBatchNorm的实现，是不是和官方模型中的BatchNorm2d不一样，可以直接在代码里面修改。maxOutput这个不要紧的，输出的是最大的置信度，至于预处理、网络结构和后处理方面确实可以讨论一下，hhh ![image](https://private-user-images.githubusercontent.com/53141838/336847155-86e3c5b3-095a-45b6-90f8-ef5a0eff8aff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2NTMxMzQsIm5iZiI6MTcxNzY1MjgzNCwicGF0aCI6Ii81MzE0MTgzOC8zMzY4NDcxNTUtODZlM2M1YjMtMDk1YS00NWI2LTkwZjgtZWY1YTBlZmY4YWZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA2VDA1NDcxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3NmVhNzlmNDkwMTk1ZjYyOGYxN2YwMjgwYzkzMTNkNWZjMDAzMzE0ZmZmYjFkYzVjNGVhZWJjM2M1YmU0YjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FqFeFaFqqCKRIcy4C6lSX__TDjDn_-F-t9Z9BAPv2WQ)

好奇怪啊，代码上用的是nn.BatchNorm2d()，怎么模型结构变成了SyncBatchNorm？
![image](https://github.com/wang-xinyu/tensorrtx/assets/167296552/60311e7a-f0c9-4360-b6d4-53329366fcdf)

**Comment by guanshanjushib** [comment]
> 你可以关注一下SyncBatchNorm的实现，是不是和官方模型中的BatchNorm2d不一样，可以直接在代码里面修改。maxOutput这个不要紧的，输出的是最大的置信度，至于预处理、网络结构和后处理方面确实可以讨论一下，hhh ![image](https://private-user-images.githubusercontent.com/53141838/336847155-86e3c5b3-095a-45b6-90f8-ef5a0eff8aff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTc2NTMxMzQsIm5iZiI6MTcxNzY1MjgzNCwicGF0aCI6Ii81MzE0MTgzOC8zMzY4NDcxNTUtODZlM2M1YjMtMDk1YS00NWI2LTkwZjgtZWY1YTBlZmY4YWZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MDYlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjA2VDA1NDcxNFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE3NmVhNzlmNDkwMTk1ZjYyOGYxN2YwMjgwYzkzMTNkNWZjMDAzMzE0ZmZmYjFkYzVjNGVhZWJjM2M1YmU0YjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.FqFeFaFqqCKRIcy4C6lSX__TDjDn_-F-t9Z9BAPv2WQ)

多卡训练时，BatchNorm2d会自动变成SyncBatchNorm，那么在生成wts的时候可以直接变成BatchNorm2d嘛，还是说只能单卡训练才能维持成BatchNorm2d层？另外SyncBatchNorm怎么实现？要是在单卡上能直接推理SyncBatchNorm层嘛

**Comment by guanshanjushib** [comment]
> 要是有兴趣，我们可以考虑实现一下SyncBatchNorm。email: [w845756@qq.com](mailto:w845756@qq.com)

可以啊，你们能实现SyncBatchNorm嘛？另外，我要是用onnx转成engine，会不会也会因为SyncBatchNorm出现精度问题？

**Comment by WuxinrongY** [comment]
> > 要是有兴趣，我们可以考虑实现一下SyncBatchNorm。email: [w845756@qq.com](mailto:w845756@qq.com)
> 
> 可以啊，你们能实现SyncBatchNorm嘛？另外，我要是用onnx转成engine，会不会也会因为SyncBatchNorm出现精度问题？
我查了一下，syncBN似乎计算上没什么区别。
另外，之前我尝试过训练yolov7的小目标，并转换engine，也有问题，原因是重参数化。

**Comment by guanshanjushib** [comment]
> > > 要是有兴趣，我们可以考虑实现一下SyncBatchNorm。email: [w845756@qq.com](mailto:w845756@qq.com)
> > 
> > 
> > 可以啊，你们能实现SyncBatchNorm嘛？另外，我要是用onnx转成engine，会不会也会因为SyncBatchNorm出现精度问题？
> > 我查了一下，syncBN似乎计算上没什么区别。
> > 另外，之前我尝试过训练yolov7的小目标，并转换engine，也有问题，原因是重参数化。

我测试了一下，wts转换出来的engine运行后，精度差异很大，用onnx转engine，运行后精度损失很少
